### PR TITLE
 stats: refine index row count estimation

### PR DESCRIPTION
**Pull request by lamxTyler**
## What have you changed? (mandatory)

Since the query pattern on the index always contains some equal conditions on the first several columns and a range condition on the last column, this pr uses CM Sketch to estimate the equal condition and histogram to estimate the range conditions, assumes that these two part are independent.

## What are the type of the changes (mandatory)?

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (mandatory)?

unit test

## Refer to a related PR or issue link (optional)

https://github.com/pingcap/tikv/pull/3206

PTAL @coocood @winoros @zz-jason @fipped

### CODE REVIEW

File: statistics/histogram.go [hunk 1]
```diff
@@ -712,7 +718,8 @@ func (c *Column) getColumnRowCount(sc *stmtctx.StatementContext, ranges []*range
 type Index struct {
 	Histogram
 	*CMSketch
-	Info *model.IndexInfo
+	statsVer int64
```
Comment (winoros) [inline statistics/histogram.go]: Add some comment explain what it is?

File: statistics/table.go [hunk 2]
```diff
@@ -44,8 +45,10 @@ type Table struct {
 	TableID     int64
 	Columns     map[int64]*Column
 	Indices     map[int64]*Index
-	Count       int64 // Total row count in a table.
-	ModifyCount int64 // Total modify count in a table.
+	colName2Idx map[string]int64 // map column name to index id
```
Comment (winoros) [inline statistics/table.go]: If we maintain histogram in plan, we will need to maintain this two map too?

File: statistics/table.go [hunk 3]
```diff
@@ -355,11 +379,88 @@ func (t *Table) GetRowCountByIndexRanges(sc *stmtctx.StatementContext, idxID int
...
+			count, err := idx.getRowCount(sc, []*ranger.Range{ran})
+			if err != nil {
+				return 0, errors.Trace(err)
+			}
+			totalCount += count
+			continue
+		}
+		selectivity := 1.0
+		// use CM Sketch to estimate the equal conditions
+		if rangePosition != 0 {
```
Comment (zz-jason) [inline statistics/table.go]: `rangePosition != 0` is guaranteed if we reach here.

File: statistics/table.go [hunk 4]
```diff
@@ -355,11 +379,88 @@ func (t *Table) GetRowCountByIndexRanges(sc *stmtctx.StatementContext, idxID int
...
+	if idx.CMSketch != nil && idx.statsVer == version1 {
+		result, err = t.getIndexRowCount(sc, idx, indexRanges)
+	} else {
+		result, err = idx.getRowCount(sc, indexRanges)
+	}
 	result *= idx.getIncreaseFactor(t.Count)
 	return result, errors.Trace(err)
 }
 
+func getRangePosition(sc *stmtctx.StatementContext, ran *ranger.Range) int {
```
Comment (zz-jason) [inline statistics/table.go]: 1. maybe `getOrdinalOfRangeCond` is a better name?
2. add some comment for this function

File: statistics/table.go (hunk 2 above)
Comment (lamxTyler) [reply statistics/table.go]: Yes, or we need to brute force over all the indices and columns...

File: statistics/selectivity_test.go [hunk 5]
```diff
@@ -20,6 +20,7 @@ import (
 	"testing"
 	"time"
 
+	"fmt"
```
Comment (shenli) [inline statistics/selectivity_test.go]: Move this to the upper part.

File: session/bootstrap.go [hunk 6]
```diff
@@ -629,6 +635,10 @@ func upgradeToVer21(s Session) {
 	doReentrantDDL(s, "ALTER TABLE mysql.gc_delete_range DROP INDEX element_id", ddl.ErrCantDropFieldOrKey)
 }
 
+func upgradeToVer22(s Session) {
+	doReentrantDDL(s, "ALTER TABLE mysql.stats_histograms ADD COLUMN `stats_ver` bigint(64) NOT NULL DEFAULT 0", infoschema.ErrColumnExists)
```
Comment (shenli) [inline session/bootstrap.go]: The `stats_ver` will be the last column. But it is not the last column in the create table statement. Please make it consistent.

**Comment by coocood** [comment]
LGTM

**Pull request merged**